### DISCUSSION TITLE
Errore creazione proforma API V2

### DISCUSSION THREAD

**Discussion by simsac81** [fattureincloud/api#158] [Q&A] [answered]
Sto provando a creare una proforma utilizzando il servizio delle API V2 https://api-v2.fattureincloud.it/c/{companyId}/issued_documents.

Di seguito il JSON inviato tramite POST

`{
  "data": {
    "id": null,
    "entity": {
      "id": 14358734,
      "code": null,
      "name": "SalvatoreXYZ",
      "type": null,
      "firstName": null,
      "lastName": null,
      "contactPerson": null,
      "vatNumber": null,
      "taxCode": null,
      "addressStreet": null,
      "addressPostalCode": null,
      "addressCity": null,
      "addressProvince": null,
      "addressExtra": null,
      "country": null,
      "email": null,
      "certifiedEmail": null,
      "phone": null,
      "fax": null,
      "notes": null,
      "defaultVat": null,
      "defaultPaymentTerms": null,
      "defaultPaymentTermsType": "STANDARD",
      "defaultPaymentMethod": null,
      "bankName": null,
      "bankIban": null,
      "bankSwiftCode": null,
      "shippingAddress": null,
      "eInvoice": null,
      "eiCode": null,
      "hasIntentDeclaration": null,
      "intentDeclarationProtocolNumber": null,
      "intentDeclarationProtocolDate": null,
      "createdAt": null,
      "updatedAt": null
    },
    "type": "PROFORMA",
    "number": null,
    "numeration": null,
    "date": null,
    "year": null,
    "currency": null,
    "language": {
      "code": "italiano",
      "name": null
    },
    "subject": "Ottobre 2022",
    "visibleSubject": "Consulenza",
    "rcCenter": "Consulenza",
    "notes": null,
    "rivalsa": null,
    "cassa": null,
    "amountCassa": null,
    "cassaTaxable": null,
    "amountCassaTaxable": null,
    "cassa2": null,
    "amountCassa2": null,
    "cassa2Taxable": null,
    "amountCassa2Taxable": null,
    "globalCassaTaxable": null,
    "amountGlobalCassaTaxable": null,
    "withholdingTax": null,
    "withholdingTaxTaxable": null,
    "otherWithholdingTax": null,
    "stampDuty": null,
    "paymentMethod": {
      "id": null,
      "name": "Pagamento a 30gg (anche lo start up)",
      "type": "STANDARD",
      "isDefault": null,
      "defaultPaymentAccount": null,
      "details": [
        {
          "title": "Intestatario",
          "description": "XYZ"
        },
        {
          "title": "IBAN",
          "description": "ITXYZEEEEEE000000000XXX"
        },
        {
          "title": "BIC/SWIFT",
          "description": "XYZ"
        }
      ],
      "bankIban": null,
      "bankName": null,
      "bankBeneficiary": null,
      "eiPaymentMethod": null
    },
    "useSplitPayment": null,
    "useGrossPrices": false,
    "eInvoice": null,
    "eiData": null,
    "eiCassaType": null,
    "eiCassa2Type": null,
    "eiWithholdingTaxCausal": null,
    "eiOtherWithholdingTaxType": null,
    "eiOtherWithholdingTaxCausal": null,
    "itemsList": [
      {
        "id": 6321162,
        "productId": null,
        "code": "D",
        "name": "Fisso mensile",
        "category": "Fisso mensile",
        "description": "Fisso mensile Ottobre 2022 Test Olga",
        "qty": 1,
        "measure": null,
        "netPrice": 100,
        "grossPrice": null,
        "vat": {
          "id": 0,
          "value": null,
          "description": null,
          "notes": null,
          "eInvoice": null,
          "eiType": null,
          "eiDescription": null,
          "editable": null,
          "isDisabled": null
        },
        "notTaxable": false,
        "applyWithholdingTaxes": null,
        "discount": null,
        "discountHighlight": null,
        "inDdt": null,
        "stock": null,
        "eiRaw": null
      },
      {
        "id": 6321183,
        "productId": null,
        "code": "I",
        "name": "Premio Hotel",
        "category": "Premio Hotel",
        "description": "10.00% incremento on line Ottobre 2022 Test Olga",
        "qty": 1,
        "measure": null,
        "netPrice": 200,
        "grossPrice": null,
        "vat": {
          "id": 0,
          "value": null,
          "description": null,
          "notes": null,
          "eInvoice": null,
          "eiType": null,
          "eiDescription": null,
          "editable": null,
          "isDisabled": null
        },
        "notTaxable": false,
        "applyWithholdingTaxes": null,
        "discount": null,
        "discountHighlight": null,
        "inDdt": null,
        "stock": null,
        "eiRaw": null
      }
    ],
    "paymentsList": [
      {
        "id": null,
        "dueDate": [
          2022,
          12,
          16
        ],
        "amount": null,
        "status": "NOT_PAID",
        "paymentAccount": null,
        "paidDate": null,
        "eiRaw": null,
        "paymentTerms": null
      }
    ],
    "template": null,
    "deliveryNoteTemplate": null,
    "accInvTemplate": null,
    "hMargins": null,
    "vMargins": null,
    "showPayments": true,
    "showPaymentMethod": true,
    "showTotals": "ALL",
    "showPaypalButton": null,
    "showNotificationButton": null,
    "showTspayButton": null,
    "deliveryNote": null,
    "accompanyingInvoice": null,
    "dnNumber": null,
    "dnDate": null,
    "dnAiPackagesNumber": null,
    "dnAiWeight": null,
    "dnAiCausal": null,
    "dnAiDestination": null,
    "dnAiTransporter": null,
    "dnAiNotes": null,
    "isMarked": null,
    "amountNet": null,
    "amountVat": null,
    "amountGross": null,
    "amountDueDiscount": null,
    "amountRivalsa": null,
    "amountRivalsaTaxable": null,
    "amountWithholdingTax": null,
    "amountWithholdingTaxTaxable": null,
    "amountOtherWithholdingTax": null,
    "amountOtherWithholdingTaxTaxable": null,
    "amountEnasarcoTaxable": null,
    "extraData": null,
    "seenDate": null,
    "nextDueDate": null,
    "url": null,
    "attachmentUrl": null,
    "attachmentToken": null,
    "eiRaw": null,
    "eiStatus": null
  },
  "options": {
    "fixPayments": true
  }
}`

Di seguito la risposta che ottengo:

`{
  "error": {
    "message": "An unexpected error occurred. Contact support for more information."
  }
}`

Il test è stato effettuato direttamente dalla pagina di test delle API.

**Comment by emajo** [comment, marked as answer]
Ciao @simsac81 è difficile capire cosa non vada in questa richiesta, prima cosa ti consiglierei di togliere tutti i campi che stai settando a null ma soprattutto di convertire i nomi dei campi da camel case in snake case, tutti i campi che stai passando in quel modo non vengono riconosciuti dalla nostra api, fatto ciò, se ottieni ancora l'errore andiamo a indagare direttamente sulla richiesta corretta.

Grazie,
Emanuele.

**Comment by simsac81** [comment]
Grazie, modificando come suggerito ho l'errore seguente:

`{
  "error": {
    "message": "Invalid request.",
    "validation_result": {
      "data.language.code": [
        "The data.language.code field is not valid."
      ]
    }
  }
}`

Attualmente ho provato a valorizzare il campo con 'IT' (da documentazione viene indicato come String quindi ipotizzavo un campo a testo libero.

Quali sono i valori ammessi per il campo language.code e language.name?

**Reply by emajo** [reply]
Passando il code a `'it'` dovrebbe risolvere il problema, per la lista di tutti i code che accettiamo è possibile fare [questa chiamata](https://developers.fattureincloud.it/api-reference#get-/info/languages).

**Comment by valmoz** [comment]
Buongiorno @simsac81 
ti consiglio di seguire [questa guida](https://developers.fattureincloud.it/docs/guides/invoice-creation) punto per punto, adattandola alle necessità della proforma.

Mauro

**Comment by simsac81** [comment]
Grazie, risolto con i vostri suggerimenti.
